[eventpipe] Use MonoImage:module_name when filename is NULL

lambdageek opened the pull request:
When assemblies are bundled (for example in WASM), the filename field is set to NULL; use the module name (from the metadata Module table).

Contributes to #72481 

With this PR and with #72859 I can generate a trace and run
```
dotnet-pgo create-mibc --trace example.nettrace -r ./bin/Debug/AppBundle/managed/'*.dll' --output /tmp/out.mibc
```

and generate a .mibc file.

lambdageek committed:
[eventpipe] Use MonoImage:name when available, if MonoImage:filename is null

When assemblies are bundled (for example in WASM), the filename field
is set to NULL and the name field is set to the basename of the
assembly.

lambdageek commented:
Sample PerfTrace line for a ModuleLoad event:

```
ThreadID="694,952" ProcessorNumber="0" ModuleID="38,139,048" AssemblyID="0" ModuleFlags="Manifest" ModuleILPath="System.Runtime.InteropServices.JavaScript.dll" ModuleNativePath="" ManagedPdbSignature="8c78d9aa-29ad-ff33-07d8-215908aea400" ManagedPdbAge="1" ManagedPdbBuildPath="/Users/alklig/work/dotnet-runtime/runtime/artifacts/obj/mono/Wasm.Browser.EventPipe.Sample/wasm/Debug/browser-wasm/linked/System.Runtime.InteropServices.JavaScript.pdb" NativePdbSignature="00000000-0000-0000-0000-000000000000" NativePdbAge="0" NativePdbBuildPath="" ModuleILFileName="System.Runtime.InteropServices.JavaScript.dll" 
```

without this PR, the ModuleILPath was `""`

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3784,7 +3784,7 @@ get_module_event_data (
 			module_data->module_flags |= MODULE_FLAGS_NATIVE_MODULE;
 
-		module_data->module_il_path = image && image->filename ? image->filename : "";
+		module_data->module_il_path = image && image->filename ? image->filename : (image->name ? image->name : "");
```
```suggestion
		module_data->module_il_path = image && image->filename ? image->filename : (image && image->name ? image->name : "");
```

lambdageek committed:
Update src/mono/mono/eventpipe/ep-rt-mono.c

radical commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3784,7 +3784,7 @@ get_module_event_data (
 			module_data->module_flags |= MODULE_FLAGS_NATIVE_MODULE;
 
-		module_data->module_il_path = image && image->filename ? image->filename : "";
+		module_data->module_il_path = image && image->filename ? image->filename : (image && image->name ? image->name : "");
```
Just curious, in what case would we get `module_il_path = ""`?

lambdageek replied:
Hm. good point.   `image->name` is always non-null. If it's not set explicitly (for example when doing `Aseembly.Load(byte[])` it's set to `"data-%p"`.  @lateralusX is it better to have `""` or some arbitrary made-up name?

lateralusX replied:
Not sure, this value can be used by tooling to locate and potentially load the module. if we have "" I guess there is bigger chance that the tool will understand that its missing, compared to having some made up value that it will probably try to use and probably fail doing.

lateralusX replied:
This is what CoreCLR does:

```
    // if we do not have a module path yet, we put the module name
    if(bIsDynamicAssembly || ModuleILPath==NULL || wcslen(ModuleILPath) <= 2)
    {
        moduleName.SetUTF8(pModule->GetSimpleName());
        ModuleILPath = (PWCHAR)moduleName.GetUnicode();
        ModuleNativePath = (PWCHAR)pEmptyString;
    }
```

So they assume they always have a valid module name and will use that as the modules IL path.

lambdageek replied:
Ok, changed it to use the module name if available and fall back to `""`

lambdageek committed:
use the module name, not image->name

module name comes from the metadata Module table and is something like
"Foo.dll" even if the assembly is loaded from a bundle or from
Assembly.Load(byte[]), whereas image->name might be "data-0x00abcdef"
for byte loads

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -3784,7 +3784,17 @@ get_module_event_data (
+			module_data->module_il_path = image->filename;
+		} else if (image && image->module_name) {
+			/* otherwise, use the module name */
+			module_data->module_il_path = imae->module_name;
```
```suggestion
			module_data->module_il_path = image->module_name;
```

lateralusX commented:
I guess the title should be Use MonoImage:filename ?

lambdageek commented:
/azp run runtime-wasm


(accidentally pushed unrelated work to this PR; reverted back to reviewed version)
